### DISCUSSION TITLE
Automatic HDL resets between multiple tests

### DISCUSSION THREAD

**Discussion by RMN-Garrison** [cocotb/cocotb#2874] [Q&A] [answered]
Hi all.
I am working on HDL designs that do not have and do not need resets (async or sync). This poses a challenge when testing them in COCOTB as the the RTL is not recompiled between tests and the HDL state of one test gets handed over to the next one! It is a requirement on my side that these tests run from a clean state, without having to start adding resets into my HDL to cater for this need.

I am interested to see if others have found this COCOTB feature of preserving HDL states between tests to be problematic and what solutions they found/would propose.

**Comment by RMN-Garrison** [comment]
Just to mention that I did explore some solutions, but the one that I believe would work best at the moment would be to use .iterate() method of the GPI handle to the simulator object to read all default values for all signals before running the test. With these values stacked up for later use, I could apply them back to the DUT before starting any subsequent tests. 
**But the .iterate() DOES NOT WORK when ran with RivieraPro simulator.** 

Help is appreciated. Thanks

**Reply by cmarqu** [reply]
Re-assigning the default values of signals statically would not take care of internal state (counters, state machines, ...).

**Reply by ktbarrett** [reply]
I agree this is not a good solution. Currently the best solution is restarting the simulator with a new test if each test must be isolated. At some point we might support reseting the simulator between each test as a form of isolation.

**Reply by RMN-Garrison** [reply]
Thanks a lot for the feedback.
Regarding the re-assignment suggestion, I would include all signals of the design. This would cover any internal states, counters etc. 
Restarting the simulator through running makefile is the nuclear option in my opinion - unfortunately does work when a TestFactory() is used.

**Comment by ktbarrett** [comment, marked as answer]
I would suggest you run each test isolated by running each test individually and then restarting the simulator. How you would accomplish this with the Makefiles:

```python
# tests.py

@cocotb.test()
async def test1(dut):
    ...

@cocotb.test()
async def test2(dut):
    ...
```

```make
# Makefile

test1:
    $(MAKE) sim MODULE=tests TESTCASE=test1

test2:
    $(MAKE) sim MODULE=tests TESTCASE=test2 

regression: test1 test2

include $(cocotb-config --makefiles)/Makefile.sim
```

**Reply by RMN-Garrison** [reply]
Do you have an idea of how I could use makefiles when TestFactory() is used to generate the actual tests?

**Reply by ktbarrett** [reply]
If you run a regression with a TestFactory, you will see it names the tests `{test_name}_001`, `{test_name}_002`, etc. You can call those names out with `TESTCASE`. As for any additional automation, there currently is none. It will be more possible in the future with the Python runner, which should accompany the 1.7 release.
